File: apps/desktop/src/lib/router/deeplink.ts
/**
 * Renderer-side deep-link handler. Pairs with `src-tauri/src/deeplink.rs`
 * which emits `deeplink:received` events with `{ kind, path, raw }`.
 */

interface DeepLinkPayload {
  kind: string;
  path: string;
  raw: string;
}

type Router = {
  navigate: (opts: { to: string; params?: Record<string, string> }) => Promise<unknown>;
};

const TAURI = typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window;

export async function installDeepLinkRouter(router: Router): Promise<() => void> {
  if (!TAURI) return () => undefined;
  const { listen } = await import('@tauri-apps/api/event');
  const unlisten = await listen<DeepLinkPayload>('deeplink:received', (event) => {
    const { kind, path } = event.payload;
    switch (kind) {
      case 'auth': {
        const url = new URL(event.payload.raw);
        const state = url.searchParams.get('state');
        const code = url.searchParams.get('code');
        if (state && code) {
          window.dispatchEvent(
            new CustomEvent('robin:auth-callback', { detail: { state, code } }),
          );
        }
        break;
      }
      case 'task': {
        const id = path.replace(/^\/+/, '');
        if (id) void router.navigate({ to: '/popover/task/$taskId', params: { taskId: id } });
        break;
      }
      case 'agent': {
        const id = path.replace(/^\/+/, '');
        if (id) void router.navigate({ to: '/expanded/agents/$agentId', params: { agentId: id } });
        break;
      }
      default:
        break;
    }
  });
  return () => unlisten();
}

File: apps/desktop/src-tauri/src/deeplink.rs
//! `robin://` deep-link routing.
//!
//! Supported URLs:
//!   robin://auth/callback?state=<state>&code=<code>
//!   robin://task/<uuid>
//!   robin://agent/<uuid>
//!
//! We hand the parsed payload to the renderer via the `deeplink:received`
//! event; the React side has a router-aware listener (`src/lib/router/deeplink.ts`).

use serde::Serialize;
use tauri::{AppHandle, Emitter, Runtime};
use tauri_plugin_deep_link::DeepLinkExt;
use url::Url;

#[derive(Debug, Clone, Serialize)]
pub struct DeepLinkPayload {
    pub kind: String,
    pub path: String,
    pub raw: String,
}

pub fn install<R: Runtime>(app: &AppHandle<R>) -> tauri::Result<()> {
    let app_handle = app.clone();
    app.deep_link().on_open_url(move |event| {
        for url in event.urls() {
            let _ = forward(&app_handle, url.as_str());
        }
    });
    Ok(())
}

pub fn handle_argv<R: Runtime>(app: &AppHandle<R>, argv: Vec<String>) {
    for arg in argv.iter() {
        if arg.starts_with("robin://") {
            let _ = forward(app, arg);
        }
    }
}

fn forward<R: Runtime>(app: &AppHandle<R>, raw: &str) -> Result<(), String> {
    let url = Url::parse(raw).map_err(|e| e.to_string())?;
    let kind = url.host_str().unwrap_or_default().to_string();
    let path = url.path().to_string();
    let payload = DeepLinkPayload {
        kind,
        path,
        raw: raw.to_string(),
    };
    app.emit("deeplink:received", &payload)
        .map_err(|e| e.to_string())?;
    Ok(())
}
